File: Algorithms/[1, 1k)/365/water-and-jug-problem.cpp
class Solution {
public:
    bool canMeasureWater(int x, int y, int z) {
        if (x+y < z) return false;
        if (!x || !y) return x == z || y == z;
        
        if (x < y) std::swap(x, y);
        x %= y;
        z %= y;
        for (int i = 0; i < y; ++i) {
            if ((i*x)%y == z) return true;
        }
        return false;
    }
};

File: Algorithms/[1, 1k)/365/water-and-jug-problem.go
package main

func canMeasureWater(x int, y int, z int) bool {
	if x+y < z {
		return false
	}
	if x*y == 0 {
		return z == x || z == y
	}
	if x < y {
		x, y = y, x
	}

	c := 0
	for {
		v := z - (x - c)
		if v/y*y == v {
			return true
		}

		c = y - (x-c)%y
		if c == 0 || c == y {
			return false
		}
	}
}
